### PR TITLE
HBASE-28850 Only return from ReplicationSink.replicationEntries while…

### PR DESCRIPTION
**Pull request by Apache9**
… all background tasks are finished

### CODE REVIEW

**Review by apurtell — COMMENTED** [review]
The trade off is this may change the performance of replicateEntries. Before, the replicateEntries call will return at the first failure. After, the replicateEntries will block for the entire time it takes to confirm all local edits in the batch are applied or failed.

I have a testbed where I reproduced SEGV crashes for HBASE-28584. Let me apply this patch there and try it out.

**Comment by Apache-HBase** [comment]
:confetti_ball: **+1 overall**






| Vote | Subsystem | Runtime |  Logfile | Comment |
|:----:|----------:|--------:|:--------:|:-------:|
| +0 :ok: |  reexec  |   0m 38s |  |  Docker mode activated.  |
| -0 :warning: |  yetus  |   0m  3s |  |  Unprocessed flag(s): --brief-report-file --spotbugs-strict-precheck --author-ignore-list --blanks-eol-ignore-file --blanks-tabs-ignore-file --quick-hadoopcheck  |
|||| _ Prechecks _ |
|||| _ master Compile Tests _ |
| +1 :green_heart: |  mvninstall  |   3m  6s |  |  master passed  |
| +1 :green_heart: |  compile  |   1m  0s |  |  master passed  |
| +1 :green_heart: |  javadoc  |   0m 31s |  |  master passed  |
| +1 :green_heart: |  shadedjars  |   5m 19s |  |  branch has no errors when building our shaded downstream artifacts.  |
|||| _ Patch Compile Tests _ |
| +1 :green_heart: |  mvninstall  |   3m  0s |  |  the patch passed  |
| +1 :green_heart: |  compile  |   1m  1s |  |  the patch passed  |
| +1 :green_heart: |  javac  |   1m  1s |  |  the patch passed  |
| +1 :green_heart: |  javadoc  |   0m 31s |  |  the patch passed  |
| +1 :green_heart: |  shadedjars  |   5m 17s |  |  patch has no errors when building our shaded downstream artifacts.  |
|||| _ Other Tests _ |
| +1 :green_heart: |  unit  | 244m 42s |  |  hbase-server in the patch passed.  |
|  |   | 269m 46s |  |  |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.47 ServerAPI=1.47 base: https://ci-hbase.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-6263/1/artifact/yetus-jdk17-hadoop3-check/output/Dockerfile |
| GITHUB PR | https://github.com/apache/hbase/pull/6263 |
| Optional Tests | javac javadoc unit compile shadedjars |
| uname | Linux cc0c19ed7609 5.4.0-195-generic #215-Ubuntu SMP Fri Aug 2 18:28:05 UTC 2024 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | dev-support/hbase-personality.sh |
| git revision | master / adc9fcce82a626cb909ab7152186b3a7dd2a9bb9 |
| Default Java | Eclipse Adoptium-17.0.11+9 |
|  Test Results | https://ci-hbase.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-6263/1/testReport/ |
| Max. process+thread count | 4570 (vs. ulimit of 30000) |
| modules | C: hbase-server U: hbase-server |
| Console output | https://ci-hbase.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-6263/1/console |
| versions | git=2.34.1 maven=3.9.8 |
| Powered by | Apache Yetus 0.15.0 https://yetus.apache.org |


This message was automatically generated.

File: hbase-server/src/main/java/org/apache/hadoop/hbase/replication/regionserver/ReplicationSink.java [hunk 1]
```diff
@@ -492,16 +492,33 @@ private void batch(TableName tableName, Collection<List<Row>> allRows, int batch
...
         if (e.getCause() instanceof TableNotFoundException) {
-          throw new TableNotFoundException("'" + tableName + "'");
+          ioe = new TableNotFoundException("'" + tableName + "'");
+        } else {
+          ioe = e;
+        }
+        if (error == null) {
+          error = ioe;
+        } else {
+          error.addSuppressed(ioe);
```
Comment (virajjasani) [inline hbase-server/src/main/java/org/apache/hadoop/hbase/replication/regionserver/ReplicationSink.java:515]: Logging the exception before it is suppressed might be helpful? (unless suppressed exceptions are all logged with original exception, which i don't recall if it happens)

**Comment by virajjasani** [comment]
Just a minor comment above, +1 otherwise. Had an offline chat with @apurtell also reg this!

**Review by apurtell — APPROVED** [review]
The change looks good. I rehydrated the testbed where I was able to reproduce SEGVs and everything has been stable and performing normally after 250M rows replicated on the way up to 1B. Not expecting any issues.

**Comment by Apache-HBase** [comment]
:confetti_ball: **+1 overall**






| Vote | Subsystem | Runtime |  Logfile | Comment |
|:----:|----------:|--------:|:--------:|:-------:|
| +0 :ok: |  reexec  |   0m 28s |  |  Docker mode activated.  |
|||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  |  No case conflicting files found.  |
| +0 :ok: |  codespell  |   0m  0s |  |  codespell was not available.  |
| +0 :ok: |  detsecrets  |   0m  0s |  |  detect-secrets was not available.  |
| +1 :green_heart: |  @author  |   0m  0s |  |  The patch does not contain any @author tags.  |
| +1 :green_heart: |  hbaseanti  |   0m  0s |  |  Patch does not have any anti-patterns.  |
|||| _ master Compile Tests _ |
| +1 :green_heart: |  mvninstall  |   2m 53s |  |  master passed  |
| +1 :green_heart: |  compile  |   3m  2s |  |  master passed  |
| +1 :green_heart: |  checkstyle  |   0m 34s |  |  master passed  |
| +1 :green_heart: |  spotbugs  |   1m 30s |  |  master passed  |
| +1 :green_heart: |  spotless  |   0m 44s |  |  branch has no errors when running spotless:check.  |
|||| _ Patch Compile Tests _ |
| +1 :green_heart: |  mvninstall  |   3m 48s |  |  the patch passed  |
| +1 :green_heart: |  compile  |   3m 28s |  |  the patch passed  |
| +1 :green_heart: |  javac  |   3m 28s |  |  the patch passed  |
| +1 :green_heart: |  blanks  |   0m  0s |  |  The patch has no blanks issues.  |
| +1 :green_heart: |  checkstyle  |   0m 45s |  |  the patch passed  |
| +1 :green_heart: |  spotbugs  |   1m 53s |  |  the patch passed  |
| +1 :green_heart: |  hadoopcheck  |  11m 59s |  |  Patch does not cause any errors with Hadoop 3.3.6 3.4.0.  |
| +1 :green_heart: |  spotless  |   0m 45s |  |  patch has no errors when running spotless:check.  |
|||| _ Other Tests _ |
| +1 :green_heart: |  asflicense  |   0m  9s |  |  The patch does not generate ASF License warnings.  |
|  |   |  38m 45s |  |  |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.43 ServerAPI=1.43 base: https://ci-hbase.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-6263/2/artifact/yetus-general-check/output/Dockerfile |
| GITHUB PR | https://github.com/apache/hbase/pull/6263 |
| Optional Tests | dupname asflicense javac spotbugs checkstyle codespell detsecrets compile hadoopcheck hbaseanti spotless |
| uname | Linux d903d5c341b7 5.4.0-1103-aws #111~18.04.1-Ubuntu SMP Tue May 23 20:04:10 UTC 2023 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | dev-support/hbase-personality.sh |
| git revision | master / adc9fcce82a626cb909ab7152186b3a7dd2a9bb9 |
| Default Java | Eclipse Adoptium-17.0.11+9 |
| Max. process+thread count | 85 (vs. ulimit of 30000) |
| modules | C: hbase-server U: hbase-server |
| Console output | https://ci-hbase.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-6263/2/console |
| versions | git=2.34.1 maven=3.9.8 spotbugs=4.7.3 |
| Powered by | Apache Yetus 0.15.0 https://yetus.apache.org |


This message was automatically generated.

**Comment by Apache-HBase** [comment]
:confetti_ball: **+1 overall**






| Vote | Subsystem | Runtime |  Logfile | Comment |
|:----:|----------:|--------:|:--------:|:-------:|
| +0 :ok: |  reexec  |   0m 38s |  |  Docker mode activated.  |
| -0 :warning: |  yetus  |   0m  4s |  |  Unprocessed flag(s): --brief-report-file --spotbugs-strict-precheck --author-ignore-list --blanks-eol-ignore-file --blanks-tabs-ignore-file --quick-hadoopcheck  |
|||| _ Prechecks _ |
|||| _ master Compile Tests _ |
| +1 :green_heart: |  mvninstall  |   3m  6s |  |  master passed  |
| +1 :green_heart: |  compile  |   0m 58s |  |  master passed  |
| +1 :green_heart: |  javadoc  |   0m 30s |  |  master passed  |
| +1 :green_heart: |  shadedjars  |   5m 15s |  |  branch has no errors when building our shaded downstream artifacts.  |
|||| _ Patch Compile Tests _ |
| +1 :green_heart: |  mvninstall  |   2m 56s |  |  the patch passed  |
| +1 :green_heart: |  compile  |   0m 58s |  |  the patch passed  |
| +1 :green_heart: |  javac  |   0m 58s |  |  the patch passed  |
| +1 :green_heart: |  javadoc  |   0m 30s |  |  the patch passed  |
| +1 :green_heart: |  shadedjars  |   5m 16s |  |  patch has no errors when building our shaded downstream artifacts.  |
|||| _ Other Tests _ |
| +1 :green_heart: |  unit  | 237m 30s |  |  hbase-server in the patch passed.  |
|  |   | 261m 46s |  |  |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.47 ServerAPI=1.47 base: https://ci-hbase.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-6263/2/artifact/yetus-jdk17-hadoop3-check/output/Dockerfile |
| GITHUB PR | https://github.com/apache/hbase/pull/6263 |
| Optional Tests | javac javadoc unit compile shadedjars |
| uname | Linux 2e2396adbb2b 5.4.0-195-generic #215-Ubuntu SMP Fri Aug 2 18:28:05 UTC 2024 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | dev-support/hbase-personality.sh |
| git revision | master / adc9fcce82a626cb909ab7152186b3a7dd2a9bb9 |
| Default Java | Eclipse Adoptium-17.0.11+9 |
|  Test Results | https://ci-hbase.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-6263/2/testReport/ |
| Max. process+thread count | 4637 (vs. ulimit of 30000) |
| modules | C: hbase-server U: hbase-server |
| Console output | https://ci-hbase.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-6263/2/console |
| versions | git=2.34.1 maven=3.9.8 |
| Powered by | Apache Yetus 0.15.0 https://yetus.apache.org |


This message was automatically generated.

File: hbase-server/src/main/java/org/apache/hadoop/hbase/replication/regionserver/ReplicationSink.java (hunk 1 above)
Comment (Apache9) [reply hbase-server/src/main/java/org/apache/hadoop/hbase/replication/regionserver/ReplicationSink.java:515]:
```
  public static void main(String[] args) {
    Exception error = new Exception("root");
    error.addSuppressed(new IOException("suppressed1"));
    error.addSuppressed(new IOException("suppressed2"));
    error.printStackTrace();
  }
```

The output
```
java.lang.Exception: root
	at org.apache.hadoop.hbase.regionserver.regionreplication.RegionReplicationSink.main(RegionReplicationSink.java:461)
	Suppressed: java.io.IOException: suppressed1
		at org.apache.hadoop.hbase.regionserver.regionreplication.RegionReplicationSink.main(RegionReplicationSink.java:462)
	Suppressed: java.io.IOException: suppressed2
		at org.apache.hadoop.hbase.regionserver.regionreplication.RegionReplicationSink.main(RegionReplicationSink.java:463)
```

So I think it is fine to not log it here?

**Pull request merged**